src/transforms/lua/v2/mod.rs

```diff
@@ -82,49 +123,208 @@ impl Lua {
+                ctx.set_named_registry_value("hooks_shutdown", Some(hooks_shutdown))?;
+            }
+
+            for (id, timer) in config.timers.iter().enumerate() {
+                let handler: rlua::Function<'_> = ctx
+                    .load(&timer.handler)
+                    .eval()
+                    .context(InvalidTimerHandler)?;
+
+                ctx.set_named_registry_value(&format!("timer_handler_{}", id), handler)?;
+                timers.push(Timer {
+                    id: id as u32,
+                    interval_seconds: timer.interval_seconds,
+                });
+            }
+
             Ok(())
-        })
-        .context(InvalidLua)?;
+        })?;
 
         Ok(Self {
             lua,
             invocations_after_gc: 0,
+            timers,
         })
     }
 
-    fn process(&mut self, event: Event) -> Result<Option<Event>, rlua::Error> {
-        let result = self.lua.context(|ctx| {
-            let globals = ctx.globals();
+    #[cfg(test)]
+    fn process(&mut self, event: Event, output: &mut Vec<Event>) -> Result<(), rlua::Error> {
+        let result = self.lua.context(|ctx: rlua::Context<'_>| {
+            ctx.scope(|scope| {
+                let emit = scope.create_function_mut(|_, event: Event| {
+                    output.push(event);
+                    Ok(())
+                })?;
+                let process = ctx.named_registry_value::<_, rlua::Function<'_>>("hooks_process")?;
+
+                process.call((event, emit))
+            })
+        });
+
+        self.attempt_gc();
+        result
+    }
 
-            globals.set("event", event)?;
+    #[cfg(test)]
+    fn process_single(&mut self, event: Event) -> Result<Option<Event>, rlua::Error> {
+        let mut out = Vec::new();
+        self.process(event, &mut out)?;
+        assert!(out.len() <= 1);
+        Ok(out.into_iter().next())
+    }
 
-            let func = ctx.named_registry_value::<_, rlua::Function<'_>>("vector_func")?;
-            func.call(())?;
-            globals.get::<_, Option<Event>>("event")
-        });
+    fn attempt_gc(&mut self) {
         self.invocations_after_gc += 1;
         if self.invocations_after_gc % GC_INTERVAL == 0 {
-            self.lua.gc_collect()?;
+            let _ = self
+                .lua
+                .gc_collect()
+                .context(RuntimeErrorGC)
+                .map_err(|e| error!(error = %e, rate_limit = 30));
             self.invocations_after_gc = 0;
         }
-        result
     }
 }
 
-impl Transform for Lua {
-    fn transform(&mut self, event: Event) -> Option<Event> {
-        match self.process(event) {
-            Ok(event) => event,
-            Err(err) => {
-                error!(message = "Error in lua script; discarding event.", error = %format_error(&err), rate_limit_secs = 30);
-                None
-            }
-        }
+impl ScriptedRuntime for Lua {
+    // TODO: make code more DRY
```

Comment:
Is this a blocker?

Reply:
It would put it into one of "code debt" milestones, as it is just about improving code quality.

Reply:
I made the implementation a bit more DRY in 6fb14e2d0cdd66e124c260cb6f8973a88035e2aa, so this TODO can be removed now.